### ISSUE TITLE
Release 0.940 planning

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#12021]
I'm planning to make a 0.940 release in February (tentative timeline). If things go according to plan, I'd cut the release branch in the first half of February.

Here are some things that I'd like to have in the release:
- #12010
- #11847
- #9636

We also need at least one typeshed sync.

Any help with the above issues would be greatly appreciated!

Please post here any issues you'd like to see fixed in the release, or PRs you'd like to have merged. After cutting the release branch, only high-priority fixes can be cherry-picked.

**Comment by AlexWaygood** [comment]
For typeshed, having the following issues fixed would be really nice, if at all possible (I'm trying to work on a solution for the first two in https://github.com/python/mypy/pull/11915):

- https://github.com/python/mypy/issues/11887
- https://github.com/python/mypy/issues/11098
- https://github.com/python/mypy/issues/11860

**Comment by cdce8p** [comment]
I would like to suggest my PR from summer last year. It's something that already works in pyright. Once mypy supports it the remaining type checkers might consider it too if they haven't already. It could help cut down duplicate overload definitions in typeshed. The PR is fully tested and only waiting for review.
* #10712

**Comment by hauntsaninja** [comment]
I have an almost green typeshed sync here: https://github.com/python/mypy/pull/11905
It's currently blocked on a single regression (which in my mind isn't a release blocker)

I'm working on a solution for #11344 that I'd really like to get in (the crash is blocking me at work / has been reported many times). I'll open an a PR in the next day or so and link it here.

Finally, https://github.com/python/mypy/issues/12005 is an issue we should fix.

**Comment by henryiii** [comment]
I'd really like to see `Self` support make it in, since it's in `typing_extensions` and supported by several other type checkers. But understandable if it isn't finished in time.

* #11666

**Comment by RJPercival** [comment]
Is there any chance of getting a v0.932 in the meantime so that we can get a fix for #11850 released? That bug has made v0.93x unusable for my project.

**Comment by intgr** [comment]
I, for one, would prefer a release of functionality that is already merged, on a shorter timescale. I would very much like to see mypy adopt time-based releases, rather than going through a feature crunch that sometimes delays useful working contributions for months.

The wait for version 0.920 was literally painful for me and my team. Now I'm waiting for support for `match` statement that was merged 2 weeks ago. It's not perfect, but it's better than the current state of even not being able to suppress warnings on `match` statements (#11829).

**Comment by sobolevn** [comment]
List of my PRs that I would like to see in this release.

## Easy

### Enum

- https://github.com/python/mypy/pull/12035
- https://github.com/python/mypy/pull/12026
- https://github.com/python/mypy/pull/11805

### Dataclasses

- https://github.com/python/mypy/pull/11496
- https://github.com/python/mypy/pull/12111
- https://github.com/python/mypy/pull/12111

### python2

- https://github.com/python/mypy/pull/11786
- https://github.com/python/mypy/pull/11787

### TypeGuard

- https://github.com/python/mypy/pull/11790
- https://github.com/python/mypy/pull/11797

## More complex ones

Most of the PRs below will change inference:
- https://github.com/python/mypy/pull/11992
- https://github.com/python/mypy/pull/11693
- https://github.com/python/mypy/pull/11128
- https://github.com/python/mypy/pull/11366
- https://github.com/python/mypy/pull/11802

Cache related:

- https://github.com/python/mypy/pull/11793

Most of them aren't reviewed yet 🙂

**Comment by hauntsaninja** [comment]
@JukkaL thoughts on release timeline? master is starting to accumulate a number of nice fixes (including some that would be appreciated at my day job)

**Comment by JukkaL** [comment]
Sorry for the delay! We've had a few(!) viruses going through our family in the last weeks. I was only able to start work on the release yesterday. Since we are late, I'm thinking of focusing on 3.10 features and low-risk changes only. This would mean postponing at least this change to the next release:
* https://github.com/python/mypy/pull/9636 (postponed)

I agree that the release process hasn't been going too smoothly in the last several releases. We can discuss options for speeding up releases separately. I've been working on some process changes that may make it feasible to switch to monthly releases, for example, but we'd probably need one or two extra volunteer release managers to help share the work.  I opened #12210 to discuss the release process.

**Comment by JukkaL** [comment]
Here's the rough priority order of my next tasks:
1. Investigate and fix/work around some regressions I've found (including https://github.com/python/typeshed/issues/7258)
2. Python 3.10 feature work (match statement, ParamSpec improvements) (time boxed)
3. Merge other fixes and improvements (time boxed)
4. Make the release!

**Comment by AlexWaygood** [comment]
#12258 is a regression on the master branch that doesn't occur on mypy 0.931.

**Comment by hauntsaninja** [comment]
Yes, will try to stay on top of it. Though the issue is actually more with "stubtest improvements on mypy master that typeshed does not yet use" than "stubtest improvements on mypy master that might not be in 0.940"

**Comment by JukkaL** [comment]
The release is now out: http://mypy-lang.blogspot.com/2022/03/mypy-0940-released.html

If you encounter or suspect a regression, please leave a comment here.

**Comment by twoertwein** [comment]
> If you encounter or suspect a regression, please leave a comment here.

Potential regression https://github.com/python/mypy/issues/12335 maybe related to the conditional overloads?

**Comment by wrobell** [comment]
Possible regression for Mypy 0.940:
```
import typing as tp
from contextlib import asynccontextmanager

T = tp.TypeVar('T')

@tp.overload
def f() -> tp.AsyncContextManager[int]:
    ...

@tp.overload
def f(cls: type[T]) -> tp.AsyncContextManager[T]:
    ...

@asynccontextmanager
async def f(cls=int):  # type: ignore
    yield 1
```
Errors:
```
$ mypy --strict tp.py
tp.py:14: error: Overloaded function implementation cannot produce return type of signature 1
tp.py:14: error: Overloaded function implementation cannot produce return type of signature 2
```
No issue with 0.931.

**Comment by hauntsaninja** [comment]
That bisects to a typeshed change: [e266cdf04f7ded2ee2da89b5912402df52a8e254] Sync typeshed (#11905)

**Comment by hauntsaninja** [comment]
Linking https://github.com/python/mypy/issues/12339 / https://github.com/python/mypy/pull/12340

**Comment by RJPercival** [comment]
This enum regression, introduced by mypy v0.930, is still present: https://github.com/python/mypy/issues/12132

**Comment by JukkaL** [comment]
Released a hotfix release 0.941 with #12340 since it seemed to affect many users. 

We'll have at least one additional bugfix release, probably either later this week or the next week.

**Comment by JukkaL** [comment]
Additional potential regressions:
* #12408 (not enough detail to reproduce)
* #12349

**Comment by JukkaL** [comment]
Mypy 0.942 is out with fixes to several regressions: https://mypy-lang.blogspot.com/2022/03/mypy-0942-released.html

At least the performance regression is still open.

**Comment by wrobell** [comment]
Please consider possible regression
```
import dataclasses as dtc
import typing as tp

T = tp.TypeVar('T')

@dtc.dataclass(frozen=True)
class S(tp.Generic[T]):
    f: tp.Callable[[T], str]

def f1(a: int) -> str:
    return str(a + 1)

def f2(a: str) -> str:
    return a + "1"

decision = {
    True: S(f1),
    False: S(f2),
}

reveal_type(decision[True])
reveal_type(decision[False])
```
Mypy 0.942 ouputs:
```
issues-mypy/t8.py:21: note: Revealed type is "builtins.object*"
issues-mypy/t8.py:22: note: Revealed type is "builtins.object*"
```
In the past the type would resolve to `S[T]`. Is the new behaviour intended?

**Comment by JelleZijlstra** [comment]
@wrobell I tried that in the playground and got the same behavior in 0.930 and as far back as mypy 0.620. I doubt it's a recent change. Which version was showing a different behavior?

**Comment by wrobell** [comment]
@JelleZijlstra I can replicate with my original codebase. Cannot replicate with the example above anymore. Not sure what's going on...

But there is still something consistently fishy. Please consider for ver. 0.942
```
import dataclasses as dtc
import typing as tp

R: tp.TypeAlias = str | int
T = tp.TypeVar('T', R, float)

@dtc.dataclass(frozen=True)
class S(tp.Generic[T]):
    v: type[T]
    f: tp.Callable[[T], str]

def f1(a: int) -> str:
    return str(a + 1)

def f2(a: str) -> str:
    return a + "1"

decision = {
    True: S(int, f1),
    False: S(str, f2),
}

reveal_type(decision[True])
reveal_type(decision[False])
```
then
```
$ mypy -V
mypy 0.942
$ mypy --strict issues-mypy/t8.py
issues-mypy/t8.py:19: error: Argument 2 to "S" has incompatible type "Callable[[int], str]"; expected "Callable[[Union[str, int]], str]"
issues-mypy/t8.py:20: error: Argument 2 to "S" has incompatible type "Callable[[str], str]"; expected "Callable[[Union[str, int]], str]"
issues-mypy/t8.py:23: note: Revealed type is "t8.S*[Union[builtins.str, builtins.int]]"
issues-mypy/t8.py:24: note: Revealed type is "t8.S*[Union[builtins.str, builtins.int]]"
Found 2 errors in 1 file (checked 1 source file)
```
When there are errors present, then the type is resolved to `S*[str | int]]`.

Let' try this:
```
import dataclasses as dtc
import typing as tp

R: tp.TypeAlias = str | int
T = tp.TypeVar('T')

@dtc.dataclass(frozen=True)
class S(tp.Generic[T]):
    v: type[T]
    f: tp.Callable[[T], str]

def f1(a: int) -> str:
    return str(a + 1)

def f2(a: str) -> str:
    return a + "1"

decision: dict[bool, S[str] | S[int]] = {
    True: S(int, f1),
    False: S(str, f2),
}

reveal_type(decision[True])
reveal_type(decision[False])
```
then
```
$ mypy --strict issues-mypy/t8.py
issues-mypy/t8.py:23: note: Revealed type is "Union[t8.S[builtins.str], t8.S[builtins.int]]"
issues-mypy/t8.py:24: note: Revealed type is "Union[t8.S[builtins.str], t8.S[builtins.int]]"
```
Setting explicit type to `decision` variable makes it work.

**Comment by hauntsaninja** [comment]
Since this isn't a regression, let's move discussion to a separate issue. The original example looks correct to me; note that if you use `T = tp.TypeVar('T', covariant=True)` you'll get `Revealed type is "__main__.S*[builtins.object]"`.

**Comment by wrobell** [comment]
Please consider this example instead
```
import dataclasses as dtc
import typing as tp

R1: tp.TypeAlias = str | int
R2: tp.TypeAlias = list | dict
T = tp.TypeVar('T', R1, R2)

@dtc.dataclass(frozen=True)
class S(tp.Generic[T]):
    f: tp.Callable[[T], str]

def f1(a: str | int) -> str:
    return str(a) + '1'

def f2(a: list | dict) -> str:
    return str(a)

#decision: dict[bool, S[str] | S[int]] = {
decision = {
    True: S(f1),
    False: S(f2),
}

reveal_type(decision[True])
reveal_type(decision[False])
```
Then for 0.931:
```
(venv) $ mypy -V
mypy 0.931
(venv) $ mypy issues-mypy/t8.py
issues-mypy/t8.py:24: note: Revealed type is "t8.S*[Any]"
issues-mypy/t8.py:25: note: Revealed type is "t8.S*[Any]"
```
and for 0.942 (0.94x actually):
```
$ mypy -V
mypy 0.942
$ mypy issues-mypy/t8.py
issues-mypy/t8.py:24: note: Revealed type is "builtins.object*"
issues-mypy/t8.py:25: note: Revealed type is "builtins.object*"
```

**Comment by hauntsaninja** [comment]
Thanks, this was because of a bad bug on my part that caused TypeAlias to work incorrectly and be treated as `Any` when not imported as `from typing import TypeAlias`: https://github.com/python/mypy/pull/12180

If you change your example to use:
```
from typing import TypeAlias

R1: TypeAlias = str | int
R2: TypeAlias = list | dict
```
you'll find that it reveals "builtins.object*" even on 0.931 (that is, modulo the TypeAlias recognition bug, the underlying type checking logic is the same).

Overall, you'll still likely find you want to use a covariant type var.

**Comment by intgr** [comment]
@wrobell If you want to report bugs, please open a separate issue and link to it -- rather than spamming the comment thread of another issue.

**Comment by wrobell** [comment]
@hauntsaninja Thanks for the explanation.

@intgr I understand where are you coming from, but such comments are not helpful.

**Comment by JelleZijlstra** [comment]
I actually agree with @intgr though I might not have used the same words. For any other regressions, please open a new separate issue and just leave a brief comment here linking back to the issue. That way, we can keep the conversation more focused.

**Comment by wrobell** [comment]
@JelleZijlstra I just followed advice in this comment https://github.com/python/mypy/issues/12021#issuecomment-1065213637. Please decide on the process and I am more than happy follow it. 🤷

**Comment by AlexWaygood** [comment]
I'm not sure if it's related to the recent release or not (it might have also happened on 0.931), but a lot of people seem to be running into #12475, which appears to be a relatively recent regression

**Comment by hauntsaninja** [comment]
Closing the issue, since it's been >1 month since 0.940 and I don't see us putting out another bugfix release. Unfortunately there doesn't yet seem to be a repro for #12475 (if it even is a regression, e.g. the original report mentions that they think it happened with 0.931 as well).
